adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m

```diff
@@ -88,79 +88,97 @@ - (void)renderNativeAdForAdConfiguration:
                  strongSelf->_nativeAd = nativeAd;
                  strongSelf->_nativeAd.delegate = strongSelf;
                  strongSelf->_nativeAd.rootViewController = adConfiguration.topViewController;
-
-                 if (strongSelf->_loadCompletionHandler) {
-                   id<GADMediationNativeAdEventDelegate> delegate =
-                       strongSelf->_loadCompletionHandler(strongSelf, nil);
-                   strongSelf->_delegate = delegate;
-                 }
+      
+                 [strongSelf loadRequireData];
                }];
 }
 
-#pragma mark - GADMediationNativeAd
-
-- (GADNativeAdImage *)icon {
-  if (!_icon) {
-    if (_nativeAd.data.icon && _nativeAd.data.icon.imageURL != nil) {
-      _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
+- (void)loadRequireData {
+    NSString *urlString = _nativeAd.data.icon.imageURL;
+    if (!urlString.length) {
+        if (_loadCompletionHandler) _delegate = _loadCompletionHandler(self,nil);
+        return;
     }
-  }
+    NSURL *url = [NSURL URLWithString:urlString];
+    if (!url) {
+        if (_loadCompletionHandler) _delegate = _loadCompletionHandler(self,nil);
+        return;
+    }
+    dispatch_async(dispatch_get_global_queue(0, 0), ^{
+        NSData *data = [NSData dataWithContentsOfURL:url];
+        dispatch_async(dispatch_get_main_queue(), ^{
+            if (!self) {
+                return;
+            }
+            GADNativeAdImage *image = [[GADNativeAdImage alloc] initWithImage:[UIImage imageWithData:data]];
+            self->_icon = image;
+            if (self->_loadCompletionHandler) {
+              id<GADMediationNativeAdEventDelegate> delegate =
+                  self->_loadCompletionHandler(self, nil);
+              self->_delegate = delegate;
+            }
+        });
+    });
```

Comment:
@wangyuanyuanyy 

Since the URL is a network URL, can you update the implementation with [dataTaskWithURL:completionHandler:](https://developer.apple.com/documentation/foundation/nsurlsession/1410330-datataskwithurl) method of the [NSURLSession](https://developer.apple.com/library/archive/technotes/tn2232/_index.html#//apple_ref/doc/uid/DTS40012884-CH1-SECNSURLSESSION) class?

Even on concurrent queues, running long synchronous methods is not ideal. Since there's a limit of 64 concurrent processes across all concurrent queues and it can fill up suprisingly quick with long sync calls.

Reply:
Do you mean the code is written like this ?
> NSURLSession *session = [NSURLSession sharedSession];
    [session dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (!self) {
                return;
            }
            GADNativeAdImage *image = [[GADNativeAdImage alloc] initWithImage:[UIImage imageWithData:data]];
            self->_icon = image;
            if (self->_loadCompletionHandler) {
              id<GADMediationNativeAdEventDelegate> delegate =
                  self->_loadCompletionHandler(self, nil);
              self->_delegate = delegate;
            }
        });
    }];

Reply:
@wangyuanyuanyy I haven't looked into the detailed implementation yet, but once you make the change to use [dataTaskWithURL:completionHandler:](https://developer.apple.com/documentation/foundation/nsurlsession/1410330-datataskwithurl) method of the [NSURLSession](https://developer.apple.com/library/archive/technotes/tn2232/_index.html#//apple_ref/doc/uid/DTS40012884-CH1-SECNSURLSESSION) class, we can review the code again.

Reply:
Ok~ I changed .